src/olympia/legacy_api/tests/test_views.py

```diff
@@ -473,11 +485,90 @@ def urlparams(x, *args, **kwargs):
+        # Number of user reviews and link to view them
+        # Total downloads, weekly downloads, and latest daily user counts
+        # Add-on creation date
+        # Link to the developer's profile
+        # File size
+        """
+        def urlparams(x, *args, **kwargs):
+            return jinja2.escape(jinja_helpers.urlparams(x, *args, **kwargs))
+
+        needles = (
+            '<addon id="4664">',
+            '<contribution_data>',
+            '<meet_developers>',
+            'https://patreon.com/blah',
+            '</meet_developers>',
+            '<reviews num="131">',
+            '%s/en-US/firefox/addon/4664/reviews/?src=api' % settings.SITE_URL,
+            '<total_downloads>1352192</total_downloads>',
+            '<weekly_downloads>13849</weekly_downloads>',
+            '<daily_users>67075</daily_users>',
+            '<author id="2519"',
+            '%s/en-US/firefox/user/cfinke/?src=api</link>' % settings.SITE_URL,
+            '<previews>',
+            'preview position="0">',
+            '<caption>TwitterBar places an icon in the address bar.</caption>',
+            'full type="image/png">',
+            '<thumbnail type="image/png">',
+            ('<developer_comments>Embrace hug love hug meow meow'
+             '</developer_comments>'),
+            'size="100352"',
+            ('<homepage>http://www.chrisfinke.com/addons/twitterbar/'
+             '</homepage>'),
+            '<support>http://www.chrisfinke.com/addons/twitterbar/</support>')
+
+        # For urls with several parameters, we need to use self.assertUrlEqual,
+        # as the parameters could be in random order. Dicts aren't ordered!
+        # We need to subtract 7 hours from the modified time since May 3, 2008
+        # is during daylight savings time.
+        url_needles = {
+            "full": urlparams(
+                '{previews}full/20/20397.png'.format(
+                    previews=jinja_helpers.user_media_url('previews')),
+                src='api', modified=1209834208 - 7 * 3600),
+            "thumbnail": urlparams(
+                '{previews}thumbs/20/20397.png'.format(
+                    previews=jinja_helpers.user_media_url('previews')),
+                src='api', modified=1209834208 - 7 * 3600),
+        }
+
+        response = make_call('addon/4664', version=1.5)
+        doc = pq(response.content)
+
+        tags = {
+            'created': ({'epoch': '1174109035'}, '2007-03-17T05:23:55Z'),
+            'last_updated': ({'epoch': '1272301783'}, '2010-04-26T17:09:43Z')}
+
+        for tag, v in tags.items():
+            attrs, text = v
+            el = doc(tag)
+            for attr, val in attrs.items():
+                assert el.attr(attr) == val
+
+            assert el.text() == text
+
+        for needle in needles:
+            self.assertContains(response, needle)
+
+        for tag, needle in url_needles.iteritems():
+            url = doc(tag).text()
+            self.assertUrlEqual(url, needle)
+
```

Comment:
This is sooooooo many `for` loops for a simple test 😿 but I am aware of all the other tests do precisely the same thing so I feel urged to go with it...

Reply:
Yeah, it's not a great test - I just copied the existing one and modified the contributions parts.  It was the safest (and easiest) route.  We can go back to forgetting the legacy API and it's tests exist again after this 😉